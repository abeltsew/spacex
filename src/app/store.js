import { configureStore } from '@reduxjs/toolkit';
import { logger } from 'redux-logger';
import missionReducer from '../feature/mission/missionSlice';
import rocketReducer from '../feature/rocket/rocketSlice';

const store = configureStore({
  reducer: {
    mission: missionReducer,
    rocket: rocketReducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(logger),
});

export default store;
