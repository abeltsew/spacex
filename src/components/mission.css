.shaded {
  background-color: rgb(183, 183, 183);
}

.table-container {
  padding: 2rem;
}

table,
th,
td {
  border: solid 1px gray;
  border-collapse: collapse;
}
