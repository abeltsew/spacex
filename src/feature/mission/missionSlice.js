import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

export const getMissions = createAsyncThunk('get/missions', async () => {
  const mission = await axios.get('https://api.spacexdata.com/v3/missions');
  return mission.data;
});
const initialState = {
  missions: [],
  isMissionLoading: true,
  missionError: undefined,
};

const missionSlice = createSlice({
  name: 'mission',
  initialState,
  extraReducers: (builder) => {
    builder.addCase(getMissions.pending, (state) => {
      state.isMissionLoading = true;
    });
    builder.addCase(getMissions.fulfilled, (state, { payload }) => {
      state.isMissionLoading = false;
      const filteredMissions = [];
      payload.map((info) =>
        filteredMissions.push({
          mission_id: info.mission_id,
          mission_name: info.mission_name,
          description: info.description,
        })
      );
      state.mission = filteredMissions;
    });
    builder.addCase(getMissions.rejected, (state, [payload]) => {
      state.missionError = payload;
    });
  },
});

export default missionSlice.reducer;
