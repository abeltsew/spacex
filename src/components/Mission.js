import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { getMissions } from '../feature/mission/missionSlice';

const Mission = () => {
  const { mission, isMissionLoading, missionError } = useSelector(
    (store) => store.mission,
  );
  const dispatch = useDispatch();

  useEffect(() => {
    if (!mission) {
      dispatch(getMissions());
    }
  }, []);

  if (isMissionLoading) {
    return <div>Loading ...</div>;
  }

  if (missionError) {
    return <div>something went wrong</div>;
  }

  return <div>{JSON.stringify(mission, null, 2)}</div>;
};

export default Mission;
