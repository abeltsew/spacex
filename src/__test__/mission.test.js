import renderer from 'react-test-renderer';
import Mission from '../components/Mission';
import { Provider } from 'react-redux';
import '@testing-library/jest-dom/extend-expect';
import configureStore from 'redux-mock-store';
import { initialState } from '../feature/mission/missionSlice';
import { render, screen } from '@testing-library/react';
import store from '../app/store';

it('should render the mission page', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Mission />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render the mission', () => {
  const mockStore = configureStore();

  render(
    <Provider store={mockStore(initialState)}>
      <Mission />
    </Provider>
  );

  const loading = screen.getByText(/Loading../i);
  expect(loading).toBeInTheDocument();
});
