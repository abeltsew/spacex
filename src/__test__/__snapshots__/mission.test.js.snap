// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the mission page 1`] = `
<div>
  Loading ...
</div>
`;
